spring-xsuaa/src/main/java/com/sap/cloud/security/xsuaa/token/TokenAuthenticationConverter.java

```diff
@@ -1,42 +1,58 @@
+/**
+ * Copyright (c) 2018 SAP SE or an SAP affiliate company. All rights reserved.
+ * This file is licensed under the Apache Software License,
+ * v. 2 except as noted otherwise in the LICENSE file
+ * https://github.com/SAP/cloud-security-xsuaa-integration/blob/master/LICENSE
+ */
 package com.sap.cloud.security.xsuaa.token;
 
-import com.sap.cloud.security.xsuaa.XsuaaServiceConfiguration;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
 import org.springframework.core.convert.converter.Converter;
 import org.springframework.security.authentication.AbstractAuthenticationToken;
 import org.springframework.security.core.GrantedAuthority;
 import org.springframework.security.core.authority.SimpleGrantedAuthority;
 import org.springframework.security.oauth2.jwt.Jwt;
 
-import java.util.Collection;
-import java.util.Collections;
-import java.util.List;
-import java.util.stream.Collectors;
+import com.sap.cloud.security.xsuaa.XsuaaServiceConfiguration;
 
+/**
+ * Converter for xsuaa jwt token that stores authorization data like scopes inside the token.
+ */
 public class TokenAuthenticationConverter implements Converter<Jwt, AbstractAuthenticationToken> {
 
-    private String appId;
+	protected String appId;
 
-    public TokenAuthenticationConverter(String appId) {
-        this.appId = appId;
-    }
+	public TokenAuthenticationConverter(String appId) {
+		this.appId = appId;
+	}
 
-    public TokenAuthenticationConverter(XsuaaServiceConfiguration xsuaaServiceConfiguration) {
-        this.appId = xsuaaServiceConfiguration.getAppId();
-    }
+	public TokenAuthenticationConverter(XsuaaServiceConfiguration xsuaaServiceConfiguration) {
+		this.appId = xsuaaServiceConfiguration.getAppId();
+	}
 
-
-    public final AbstractAuthenticationToken convert(Jwt jwt) {
-        Collection<GrantedAuthority> authorities = extractAuthorities(jwt);
-        return new AuthenticationToken(appId, jwt, authorities);
-    }
+	public final AbstractAuthenticationToken convert(Jwt jwt) {
+		return new AuthenticationToken(appId, jwt, extractAuthorities(jwt));
+	}
 
     protected Collection<GrantedAuthority> extractAuthorities(Jwt jwt) {
-        Collection<String> scopes = this.getScopes(jwt);
-        return scopes.stream().map(authority -> authority).map(SimpleGrantedAuthority::new).collect(Collectors.toList());
-    }
+        Collection<String> scopeAuthorities = getScopes(jwt);
+        Collection<String> otherAuthorities = extractAuthorities(new TokenImpl(jwt, appId));
```

Comment:
Libraries only validate the scope field.
All user tokens (password grant, SAML, authorization grant) only contain scopes, the client credential tokens contain scopes and authorities. Scopes are filtered by using /oauth/token?scope=... but authorities are not.
Therefore validation must only look into the scope field.

Reply:
No, the extractAuthorities method returns an empty collection:
https://github.com/SAP/cloud-security-xsuaa-integration/blob/d6e71ea7944d1e08fc538158497203c1b83c7fa9/spring-xsuaa/src/main/java/com/sap/cloud/security/xsuaa/token/TokenAuthenticationConverter.java#L51